main.go

```diff
@@ -607,13 +610,53 @@ func main() {
 			internalserver.WithPProf(),
 		)
 
+		internalTLSConfig, err := tls.NewServerConfig(
+			log.With(logger, "protocol", "HTTP"),
+			cfg.tls.internalServerCertFile,
+			cfg.tls.internalServerKeyFile,
+			cfg.tls.minVersion,
+			cfg.tls.cipherSuites,
+		)
+		if err != nil {
+			stdlog.Fatalf("failed to initialize tls config: %v", err)
+		}
+
+		if internalTLSConfig != nil {
+			r, err := rbacproxytls.NewCertReloader(
+				cfg.tls.internalServerCertFile,
+				cfg.tls.internalServerKeyFile,
+				cfg.tls.reloadInterval,
+			)
+			if err != nil {
+				stdlog.Fatalf("failed to initialize certificate reloader: %v", err)
+			}
+
+			internalTLSConfig.GetCertificate = r.GetCertificate
+
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				return r.Watch(ctx)
+			}, func(error) {
+				cancel()
+			})
+		}
+
 		s := http.Server{
-			Addr:    cfg.server.listenInternal,
-			Handler: h,
+			Addr:         cfg.server.listenInternal,
+			Handler:      h,
+			TLSConfig:    internalTLSConfig,
+			ReadTimeout:  readTimeout,  // best set per handler.
+			WriteTimeout: writeTimeout, // best set per handler.
 		}
 
 		g.Add(func() error {
 			level.Info(logger).Log("msg", "starting internal HTTP server", "address", s.Addr)
+
+			if internalTLSConfig != nil {
+				// serverCertFile and serverKeyFile passed in TLSConfig at initialization.
```

Comment:
nit: I guess if we'd like to retain the comment
```suggestion
				// internalServerCertFile and internalServerKeyFile passed in TLSConfig at initialization.
```